### ISSUE TITLE
spring-rabbit-test: rabbitListenerAnnotationProcessor has already been defined

### ISSUE THREAD

**Issue by liias** [spring-projects/spring-amqp#914]
I am using:
spring boot 2.1.2.RELEASE
spring-rabbit-2.1.3.RELEASE
spring-rabbit-test-2.1.3.RELEASE

When running tests I get an error:

`The bean 'org.springframework.amqp.rabbit.config.internalRabbitListenerAnnotationProcessor', defined in class path resource [org/springframework/amqp/rabbit/test/RabbitListenerTestBootstrap.class], could not be registered. A bean with that name has already been defined in class path resource [org/springframework/amqp/rabbit/annotation/RabbitBootstrapConfiguration.class] and overriding is disabled.`

Spring suggests to use "spring.main.allow-bean-definition-overriding=true" to override the default behaviour and it does work. However, I find it a bit bad solution to change global spring configuration.

There is also possibility that I am doing something wrong.

**Comment by liias** [comment]
Just looked into how **RabbitBootstrapConfiguration** gets loaded in my tests.

Using @RabbitListenerTest does two things:
Enables @EnableRabbit (which does @Import(**RabbitBootstrapConfiguration**.class))
Adds @Import(RabbitListenerTestBootstrap.class)

Now, I can't do @Import(RabbitListenerTestBootstrap.class) instead of @RabbitListenerTest, because 
rabbit-test checks the annotation specifically in org.springframework.amqp.rabbit.test.RabbitListenerTestHarness#RabbitListenerTestHarness.

And even if I could, there are probably some parts of RabbitBootstrapConfiguration I do want?

**Comment by garyrussell** [comment]
I am not sure why you opened an issue here; this has nothing to do with this project.

Can you explain what you think we can do about it?

**Comment by liias** [comment]
Hey.

So, I don't know the project well to state facts, so here are my guesses:
You could take away `@EnableRabbit` from `@RabbitListenerTest`, so RabbitBootstrapConfiguration is not loaded in. In order to get other beans from RabbitBootstrapConfiguration, the RabbitBootstrapConfiguration could be split up or you could redefine all beans in RabbitListenerTestBootstrap, so RabbitListenerTestBootstrap would be a full replacement for RabbitBootstrapConfiguration.

Does that make any sense?

**Comment by garyrussell** [comment]
It makes sense for Boot users but it will break every application that uses `@RabbitListenerTest` but not Spring Boot.

I don't know if it's possible to make it conditional on whether Boot is present on the class path, but it could be difficult because we'd end up with circular dependencies.

**Comment by liias** [comment]
If nothing else, then this issue will at least appear to people starting a new project with the test lib.

So you are saying spring.main.allow-bean-definition-overriding=false is very much a spring-boot default, and won't make it into spring _proper_ as default value?

**Comment by garyrussell** [comment]
The problem is that if we, say, unconditionally set `spring.main.allow-bean-definition-overriding=true` in the test framework, it could mask the very issue that the Boot team is trying to avoid by making it false by default.

What is the difficulty of adding a `TestPropertySource` to your test case?

```java
@RabbitListenerTest
@TestPropertySource(properties = "spring.main.allow-bean-definition-overriding=true")
```

**Comment by liias** [comment]
`@TestPropertySource` certainly works. But I would personally just enable it in my application, because if just enabling it when running tests you could inadvertently get errors when actually running application (because tests passed).

In conclusion, I thought spring is advertising to not override beans now and this library (which is part of spring umbrella) does do that, so wanted to let you know.

**Comment by garyrussell** [comment]
Good point; we'll take a look to see if there is something we can do to avoid this.

**Comment by garyrussell** [comment]
Fixed in 2.1.5.BUILD-SNAPSHOT

**Comment by garyrussell** [comment]
Fixed in 2.1.5.BUILD-SNAPSHOT - thanks for persevering!

**Comment by liias** [comment]
This was the fastest issue to fix time I've seen for third party library. High five to both of you.
